technical/hashmap/c++/hashmap.h

```diff
@@ -0,0 +1,131 @@
+#include <vector>
+#include <string>
+
+static const size_t MIN_CAPACITY = 4;
+static const double UPPER_LOAD_FACTOR = 0.75;
+static const double LOWER_LOAD_FACTOR= 0.25;
+
+template <typename Value>
+class HashMap {
+private:
+  struct bucket {
+    std::string key;
+    Value value;
+    bucket() {}
+    bucket(const std::string& k, const Value& v): key(k), value(v) {}
+  };
+  bucket* buckets;
+  size_t len;
+  size_t capacity;
+  double load;
+  bool* valid;
+  std::hash<std::string> hasher;
+  // compressess a hashvalue to the space [0, N-1]
+  size_t compressor(size_t hashvalue) const {return hashvalue % capacity;}
+  // update the load factor for the table
+  void updateLoad() {
+    load = (double)len / (double)capacity;
+  }
+  // grow the hashtable and rehash all of the items to the new table.
+  // expensive operation with O(N) runtime and O(N) space
+  void rehashTo(size_t next_capacity) {
+    size_t old_capacity = capacity;
+    capacity = next_capacity;
+    bucket* old_buckets = buckets; // save current buckets
+    bool* old_valid = valid;
+    // blank slate buckets with next_capacity space
+    buckets = new bucket[next_capacity];
+    // blank slate valid with next_capacity space
+    valid = new bool[next_capacity];
+    for(int i = 0; i < capacity; i++) {
+      valid[i] = true;
+    }
+    load = 0; // reset load
+    // insert everything from old_buckets into new
+    for(int i = 0; i < old_capacity; i++) {
+      if(!old_valid[i]) {
+        put(old_buckets[i].key, old_buckets[i].value);
+      }
+    }
```

Comment:
Can we check for the updated load, even though it should still be between the limits